keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,327 @@
+
+### User Stories
+
+- As a cluster user, I want to create a secret that carries the private keys required for my encrypted images
+- As a cluster user, I want to run the encrypted container images using the private keys carried by ImageDecryptSecret
+- As a cluster operator, I want to add the ImageDecryptSecret to the service account
+- As a cluster user, I want to run the encrypted images using the private keys from the service account
+- As an application developer, I want to encrypt my container images and be able to run them securely using kubernetes
+- As an application developer, I want to protect the content of my container images such that only me (as an application developer) and the execution runtime can read them. Any other third party, such as container registry, should **NOT** be able to read the content of my container image.
+
+## Proposal
+
+The initial design includes:
+
+- `ImageDecryptSecret` API resource definition
+- `ImageDecryptSecret` pod field for specifying the ImageDecryptSecret the pod should be run with
+- Kubelet implementation for fetching & interpreting the ImageDecryptSecret
+- CRI API & implementation for passing along the ImageDecryptSecret
+
+### API
+
+`ImageDecryptSecret` is a new cluster-scoped Secret
+
+
+The private key is selected by the pod by specifying the ImageDecryptSecret in the PodSpec. Once the pod is
+scheduled, the ImageDecryptSecret cannot be changed.
+
+_(This is a simplified declaration, syntactic details will be covered in the API PR review)_
+
+```go
+// Container represents a single container that is expected to be run on the host.
+type Container struct {
+  <snip>
+  // ImageDecryptSecrets is an optional list of references to secrets in the same namespace to use for decrypting any of the images used by this PodSpec.
+  // If specified, these secrets will be passed to individual puller implementations for them to use.
+  // +optional
+  ImageDecryptSecrets []LocalObjectReference
+  </snip>
+}
+```
+[Reference Implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/pkg/apis/core/types.go#L1915) of the `Container` struct
+
+```go
+const (
+  <snip>
+  // SecretTypeDecryptKeys defines the type for the decrypt secrets
+  SecretTypeDecryptKey SecretType = "kubernetes.io/decryptionkey"
+
+  // ImageDecryptionKeys represent the key required to access secret data
+  ImageDecryptionKey = ".imagedecryptionkey"
+  </snip>
+)
+```
+[Reference Implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/staging/src/k8s.io/api/core/v1/types.go#L4986) for the `constant` that hold the actual private key required for decrypting the image
+
+
+In order to support ImageDecryptSecret using service account, we need to define a struct that holds the ImageDecryptSecret and the corresponding image(s) that can be decrypted.
```

Comment:
Can you explain why the image list is needed? What is it used for? It seems not covered in the doc.

And it is hard to maintain an exhausted list of images, if the key is used to encrypt a lot of images.

Reply:
In the initial design, there was a strong coupling between the `ImageDecryptSecret` and the container image. This was done so that the private keys from the given `ImageDecryptSecret` are sent to runtime for image decryption only for those images that are found in that list. For any other container image that's not encrypted, the implementation of image puller in CRI would have not gotten the decryption keys. 

But I see your point of having to maintain the list of image, also I realized that by not sending the keys for unencrypted images we are just saving an insignificant amount of data in the payload of CRI call to pull the image. 

So considering all this, I have removed the need to maintain the list of images in service account.

Reply:
SG.